otherlibs/systhreads/decs.c

```diff
@@ -0,0 +1,899 @@
+  #endif
+}
+
+static void caml_dec_enter_blocking_section(void)
+{
+  Current_dec = pthread_getspecific(Dec_key);
+  caml_dec_save_runtime_state();
+  pthread_setspecific(Dec_key, Current_dec);
+  caml_dec_main_lock_release(&Dec_main_lock);
+}
+
+static void caml_dec_leave_blocking_section(void)
+{
+  caml_dec_main_lock_acquire(&Dec_main_lock);
+  Current_dec = pthread_getspecific(Dec_key);
+  caml_dec_restore_runtime_state();
+}
+
+static caml_dec_t caml_dec_new_info(void)
+{
+  caml_dec_t dec;
+  struct domain *d;
+
+  d = caml_domain_self();
+  dec = NULL;
+  dec = (caml_dec_t) caml_stat_alloc_noexc(sizeof(struct caml_dec_struct));
+  if (dec == NULL) return NULL;
+
+  dec->descr = Val_unit;
+  dec->current_stack = caml_alloc_main_stack(Stack_size / sizeof(value));;
+  dec->c_stack = NULL;
+  dec->exn_handler = NULL;
+  dec->local_roots = NULL;
+  dec->exit_buf = NULL;
+  dec->backtrace_pos = 0;
+  dec->backtrace_buffer = NULL;
+  dec->backtrace_last_exn = caml_create_root(Val_unit);
+  dec->domain_id = d->state->id;
+
+  #ifndef NATIVE_CODE
+  dec->trap_sp_off = 1;
+  dec->trap_barrier_off = 2;
+  dec->external_raise = NULL;
+  #endif
+
+  return dec;
+}
+
+static value caml_dec_new_descriptor(value clos)
+{
+  CAMLparam1(clos);
+  CAMLlocal2(descr, mu);
+
+  mu = Val_unit;
+  mu = caml_decstatus_new(Val_unit);
+  descr = caml_alloc_small(3, 0);
+  Ident(descr) = Val_long(atomic_load_acq(&dec_next_id));
+  Clos(descr) = clos;
+  Terminated(descr) = mu;
+  atomic_fetch_add(&dec_next_id, +1);
+
+  CAMLreturn (descr);
+}
+
+// used to clean context after fork.
+static void caml_dec_reinitialize(void)
+{
+  caml_dec_t dec, next;
+
+  dec = Current_dec->next;
+  while (dec != Current_dec) {
+    next = dec->next;
+    caml_stat_free(dec);
+    dec = next;
+  }
+  Current_dec->next = Current_dec;
+  Current_dec->prev = Current_dec;
+  All_decs = Current_dec;
+
+  // within the child, the domain_lock needs to be reset
+  // and acquired.
+  caml_reset_domain_lock();
+  caml_bt_acquire_domain_lock();
+  // main_lock needs to be initialized and released.
+  caml_dec_main_lock_init(&Dec_main_lock);
+  caml_dec_main_lock_release(&Dec_main_lock);
+}
+
+static int dec_atfork(void (*fn)(void))
+{
+  return pthread_atfork(NULL, NULL, fn);
+}
+
+CAMLprim value caml_dec_initialize(value unit);
+
+static void caml_dec_domain_start_hook(void) {
+  caml_dec_initialize(Val_unit);
+}
+
+CAMLprim value caml_dec_initialize(value unit)
+{
+  CAMLparam0();
+
+  caml_dec_t new_dec;
+
+  caml_dec_main_lock_init(&Dec_main_lock);
+
+  new_dec =
+    (caml_dec_t) caml_stat_alloc_noexc(sizeof(struct caml_dec_struct));
+
+  new_dec->descr = caml_dec_new_descriptor(Val_unit);
+  new_dec->next = new_dec;
+  new_dec->prev = new_dec;
+
+  #ifdef NATIVE_CODE
+  new_dec->exit_buf = &caml_termination_jmpbuf;
+  #endif
+
+  // Hooks setup, if caml_scan_roots_hook is set, it was done already
+  // by another domain.
+  if (caml_scan_roots_hook != caml_dec_scan_roots) {
```

Comment:
I don't think this is going to be safe if two domains try to initialize at the same time: 
- domain A starts initializing and gets `caml_scan_roots_hook` set to `caml_dec_scan_roots`, but gets suspended before the other hooks are initialized
- the prev_scan_roots_hook gets written with something unexpected

Could put in a mutex or there you could way with a CAS on `prev_scan_roots_hook` and be careful with the ordering of the operations.